cmd/kubeadm/app/componentconfigs/configset.go

```diff
@@ -0,0 +1,204 @@
+		return runtime.DecodeInto(Codecs.UniversalDecoder(), yaml, into)
+	}
+
+	return nil
+}
+
+// known holds the known component config handlers. Add new component configs here.
+var known = []*handler{
+	&kubeProxyHandler,
+	&kubeletHandler,
+}
+
+// ensureInitializedComponentConfigs is an utility func to initialize the ComponentConfigMap in ClusterConfiguration prior to possible writes to it
+func ensureInitializedComponentConfigs(clusterCfg *kubeadmapi.ClusterConfiguration) {
+	if clusterCfg.ComponentConfigs == nil {
+		clusterCfg.ComponentConfigs = kubeadmapi.ComponentConfigMap{}
+	}
+}
+
+// Default sets up defaulted component configs in the supplied ClusterConfiguration
+func Default(clusterCfg *kubeadmapi.ClusterConfiguration, localAPIEndpoint *kubeadmapi.APIEndpoint) {
+	ensureInitializedComponentConfigs(clusterCfg)
+
+	for _, handler := range known {
+		// If the component config exists, simply default it. Otherwise, create it before defaulting.
+		group := handler.GroupVersion.Group
+		if componentCfg, ok := clusterCfg.ComponentConfigs[group]; ok {
+			componentCfg.Default(clusterCfg, localAPIEndpoint)
+		} else {
+			componentCfg := handler.CreateEmpty()
+			componentCfg.Default(clusterCfg, localAPIEndpoint)
+			clusterCfg.ComponentConfigs[group] = componentCfg
+		}
+	}
+}
+
+// FetchFromCluster attempts to fetch all known component configs from their config maps and store them in the supplied ClusterConfiguration
+func FetchFromCluster(clusterCfg *kubeadmapi.ClusterConfiguration, client clientset.Interface) error {
+	ensureInitializedComponentConfigs(clusterCfg)
+
+	for _, handler := range known {
+		componentCfg, err := handler.FromCluster(client, clusterCfg)
+		if err != nil {
+			return err
+		}
+
+		if componentCfg != nil {
+			clusterCfg.ComponentConfigs[handler.GroupVersion.Group] = componentCfg
+		}
+	}
+
+	return nil
+}
+
+// FetchFromDocumentMap attempts to load all known component configs from a document map into the supplied ClusterConfiguration
+func FetchFromDocumentMap(clusterCfg *kubeadmapi.ClusterConfiguration, docmap kubeadmapi.DocumentMap) error {
+	ensureInitializedComponentConfigs(clusterCfg)
+
+	for _, handler := range known {
+		componentCfg, err := handler.FromDocumentMap(docmap)
+		if err != nil {
+			return err
+		}
+
+		if componentCfg != nil {
+			clusterCfg.ComponentConfigs[handler.GroupVersion.Group] = componentCfg
+		}
+	}
+
+	return nil
+}
+
+// Validate is a placeholder for performing a validation on an already loaded component configs in a ClusterConfiguration
+// Currently it prints a warning that no validation was performed
+func Validate(clusterCfg *kubeadmapi.ClusterConfiguration) field.ErrorList {
+	groups := []string{}
+	for group := range clusterCfg.ComponentConfigs {
+		groups = append(groups, group)
+	}
+	sort.Strings(groups) // The sort is needed to make the output predictable
+	klog.Warningf("WARNING: kubeadm cannot validate component configs for API groups %v", groups)
```

Comment:
i think we saw this warning being printed in a lot of commands such as `kubeadm config images`.
should we change it to klog.V(...).Infof()?

Reply:
I guess so. The code is executed whenever the config is loaded (which happens in almost all kubeadm commands). However it's good to see it as a warning only on `kubeadm init/upgrade *`.